src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -839,23 +810,27 @@ public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync
             {
                 if (sessionChannelDispatcher == null)
                 {
-                    if (!this.settings.RemovePendingSession(this.sessionToken.ContextId))
-                    {
-                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new CommunicationException(SR.Format(SR.SecuritySessionNotPending, this.sessionToken.ContextId)));
-                    }
-                    if (this.settings.AcceptorChannelType is IDuplexChannel)
-                    {
-                        //TODO later
-                    }
-                    else
+                    lock (this.settings.ThisLock)
                     {
-                        ServerSecuritySimplexSessionChannel.SecurityReplySessionServiceChannelDispatcher
-                            replySessionChannelDispatcher = new ServerSecuritySimplexSessionChannel.
-                            SecurityReplySessionServiceChannelDispatcher(this.settings, this.sessionToken,
-                            null, this.settings.SettingsLifetimeManager, channel, this.remoteAddress);
-                        await replySessionChannelDispatcher.OpenAsync(ServiceDefaults.OpenTimeout);
-                        sessionChannelDispatcher = replySessionChannelDispatcher;
-                        settings.AddSessionChannel(this.sessionToken.ContextId, replySessionChannelDispatcher, this.messageFilter);
+                        if (!this.settings.RemovePendingSession(this.sessionToken.ContextId))
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new CommunicationException(SR.Format(SR.SecuritySessionNotPending, this.sessionToken.ContextId)));
+                        }
+                        if (this.settings.AcceptorChannelType is IDuplexChannel)
+                        {
+                            //TODO later
+                        }
+                        else
+                        {
+                            ServerSecuritySimplexSessionChannel.SecurityReplySessionServiceChannelDispatcher
+                                replySessionChannelDispatcher = new ServerSecuritySimplexSessionChannel.
+                                SecurityReplySessionServiceChannelDispatcher(this.settings, this.sessionToken,
+                                null, this.settings.SettingsLifetimeManager, channel, this.remoteAddress);
+                            Task openTask = replySessionChannelDispatcher.OpenAsync(ServiceDefaults.OpenTimeout);
```

Comment:
Does this need to happen under lock? I'm not seeing any common state that it getting touched here other than the AddSessionChannel which takes the lock anyway.

Reply:
out of caution.. think about from the client side once the channel opened, client is sending concurrent requests(using same token) under same channel.. in that case we might open up 2 or more service channel dispatcher... let me know what your thought..